[Refactor] Installer v2 with support for backup

kylo252 opened the pull request:
# Description

- Provide an install script that is more coherent and doesn't require root privileges. Instead it will prompt the user to install any missing requirements with a hint about the possible command.
- Add built-in support for backing up any existing lunarvim folders.
- Add support for cloning from a fork `LV_REMOTE=..`
- Add support for overriding all the destination folders `$LUNARVIM_RUNTIME_DIR`, `$LUNARVIM_CONFIG_DIR`, `$INSTALL_PREFIX`
- Add support for updating automatically if it was already installed and there are no conflicts (--ff-only)

## How Has This Been Tested?

```shell
wget https://raw.githubusercontent.com/kylo252/LunarVim/installer_v2/utils/installer/install.sh
LVBRANCH=installer_v2 LV_REMOTE="kylo252/lunarvim.git" INSTALL_PREFIX=/tmp/t1 LUNARVIM_CONFIG_DIR=/tmp/t1 LUNARVIM_RUNTIME_DIR=/tmp/t2 bash ./install.sh
```

abzcoding commented on utils/installer/install.sh:
```diff
@@ -1,225 +1,191 @@
+        because of the --overwrite flag"
+      read -p "Would you like to continue? [y]es or [n]o : " -r answer
+      [ "$answer" == "${answer#[Yy]}" ] && exit 1
+      for dir in "${__lvim_dirs[@]}"; do
```
we might need to remove `NEOVIM_CACHE_DIR` as well (`~/.cache/nvim`)

kylo252 replied:
Yeah but if it's holding the sessions_dir then that will affect people's neovim's setup. So I'm not sure how to gracefully do this.

abzcoding replied:
this is the `--overwrite` flag so the user has asked for it :smiling_imp:

devtoi commented on utils/installer/install.sh:
```diff
@@ -1,225 +1,191 @@
+
+  if [ -f "$LUNARVIM_RUNTIME_DIR/lvim/utils/bin/lvim" ]; then
+    echo "Install LunarVim shim"
+    cp "$LUNARVIM_RUNTIME_DIR/lvim/utils/bin/lvim" "$HOME/.local/bin"
```
Can we expect that the user had this path in the PATH?

kylo252 replied:
So I'm relying on the fact that people should have this, since it's used by `pip` for example. 

I'm also considering removing this all together and use a shell function and append it to the user bashrc/zshrc. But then we'd need to ask the user every time.

devtoi replied:
I think the current one in this PR is the lesser evil. Changing user configs are gonna be iffy at best, imo.

devtoi commented on utils/installer/install.sh:
```diff
@@ -1,225 +1,191 @@
+  if [ -e "$LUNARVIM_RUNTIME_DIR/lvim/init.lua" ]; then
+    echo "LunarVim already installed"
+    echo "checking for the latest updates instead"
+    local CMD="git -C $LUNARVIM_RUNTIME_DIR/lvim pull origin $LVBRANCH"
```
Should this case perhaps complain if the existing remote doesn't match the expected?

kylo252 replied:
You mean if it's not called "origin" right? damn that got quite a bit more complicated just now. I should create a function for it which will actually prompt the user before doing anything to the repo.

devtoi replied:
Yeah, or if the supplied LV_REPO isn't the same as origin

kylo252 replied:
Yeah this is way too complicated. I'm gonna leave this for a separate PR with regards on how to handle updating lunarvim. 
There are simply way too many possible cases so it's not trivial. I guess we could open a poll and see what people think the workflow should look like.

abzcoding commented:
also we need to check if `git` is available or not
(adding `git` and `neovim` as dependencies )

abzcoding commented on utils/installer/install.sh:
```diff
@@ -1,225 +1,191 @@
+      cp -r "$(pwd)" "$LUNARVIM_RUNTIME_DIR/lvim"
+      ;;
+    *)
+      local CLONE_CMD="git clone https://github.com/$LV_REPO --branch $LVBRANCH --depth 1 \
```
apparently `--depth 1` caused some issues.
cc: @rebuilt

kylo252 replied:
> also we need to check if `git` is available or not
> (adding `git` and `neovim` as dependencies )

I had a system_dep check that I can bring back for this.

abzcoding commented on utils/installer/install.sh:
```diff
@@ -1,225 +1,191 @@
+function clone_lvim() {
+  echo "Backing up old LunarVim configuration"
+  backup_old_config
+  echo "Cloning LunarVim configuration"
```
on fresh install do we need to check for the existence of `LUNARVIM_RUNTIME_DIR` ?

kylo252 replied:
That's actually already handled by the backup function. I took special note to keep that logic.

abzcoding commented on .github/workflows/format.yaml:
```diff
@@ -39,5 +39,5 @@ jobs:
       - name: Check formatting
         run: |
-          shfmt -l -d .
+          shfmt -i 2 -ci -l -d .
```
cc: @ChristianChiarulli

abzcoding replied:
@kylo252 if we are changing this, we might wanna change the default formatting on `core/lang/sh.lua` as well
https://github.com/ChristianChiarulli/LunarVim/blob/0064b446a02be13dba0f16f084e073b347716710/lua/lang/sh.lua#L13-L17

kylo252 replied:
I'm gonna move that shfmt style commit to a separate PR since it's affecting a few files already.

abzcoding replied:
sounds good to me :+1:

sperezconesa commented:
Some feedback installing in Ubuntu:
+ The script asks to `sudo apt install -y  nvim` it should be `neovim` I think.
+ Somewhere it is written "noevim".
+ I get this when installing:
![image](https://user-images.githubusercontent.com/27727343/129183298-a64da8ee-fa80-4f7d-a07f-1e381899e9e7.png)

kylo252 commented:
> Some feedback installing in Ubuntu:
> 
> * The script asks to `sudo apt install -y  nvim` it should be `neovim` I think.
> * Somewhere it is written "noevim".
> * I get this when installing:

I rebased the script and added update functionality (when safe), can you try again please?

sperezconesa commented:
Installation works now! But if I try to start lvim in:
```bash
.local/bin/lvim
./test_lvim/bin/lvim
```
I just get regular plain neovim. I guess environment variables are missing.

kylo252 commented:
> Installation works now! But if I try to start lvim in:
> ```bash
> .local/bin/lvim
> ./test_lvim/bin/lvim
> ```
> I just get regular plain neovim. I guess environment variables are missing.

I forgot to make the shim to be dynamic as well. Any ideas how to handle that, other than writing a new file?

abzcoding commented:
@kylo252 
can't we install all the dependencies if the `--testing` flag is given, or fix the installation github workflow

Tastyep reviewed (changes requested):
Looks neat :+1: 

My only complaint is that we now require fd-find, I don't have it, am I missing out on something ?

kylo252 commented:
> Looks neat 👍
> 
> My only complaint is that we now require fd-find, I don't have it, am I missing out on something ?

I'm surprised it's not the default one in Telescope. You should definitely install it if you don't have it already. I'm also considering dropping it along with ripgrep from this table.

kylo252 commented on utils/installer/install.sh:
```diff
@@ -1,272 +1,271 @@
+  if [ ! -d "$INSTALL_PREFIX/bin" ]; then
+    mkdir -p "$INSTALL_PREFIX/bin"
   fi
+  cat >"$INSTALL_PREFIX/bin/lvim" <<EOF
```
I would generally agree, but try to write that `sed` command in a way that doesn't expand variables. I hate it when I have a hard-coded expansion of `$HOME` in my scripts, and I would like to keep those variables. That's why I opted for this approach which is way easier to read/maintain, but it's also very possible that there's a nice way to do it with `sed`.

Some notes to keep in mind about making sure it's cross-platform:
- `sed -E` is different between Linux and Darwin
- `awk` isn't guaranteed to be available and/or work reliably on BSD

time to write it in `perl`? 😄

Tastyep replied:
I don't understand what you mean about the sed command expanding variables, that's the whole point.
Please explain to me how having two versions of the same file, `bin/lvim` and this one hidden in the setup script is more maintainable.

kylo252 replied:
We don't need to have `util/bin/lvim` as a static file anymore. And there's nothing to maintain that isn't handled by the installer.

Tastyep replied:
What am i missing here ?
https://github.com/LunarVim/LunarVim/pull/1052/files#diff-fe4492969d1077edb685fdb1cc8cf4e6a7d93dcf6eb20d6b4e8fdd908077a5fa

kylo252 replied:
I meant I'm gonna delete it in the next patch 😄

Tastyep replied:
Sorry to come back on the topic, but I still think this should not be generated from the script.
These lines are unnecessarily hidden in the setup script, what didn't you like about the solution I offered ?

kylo252 replied:
It's does not qualify as "hidden" because: 
1- It's not doing anything out of the ordinary, as `nvim -u` is a very common flag.
2- the binary is entirely dependent on the install script, and there is not a single useful thing that it does which is not controlled by the installer. 

This pattern is not as uncommon as you think it is, especially for a trivial shim. You can also see similar patterns in homebrew or nix for example. 

In a later version, the user will be prompted to add `lvim` as a shell alias/function, that will also help in maintaining autocompletion. or allow running as a default neovim config by adjusting some variables and adding a symlink to `vim.fn.stdpath "config"`

Tastyep replied:
I've given my arguments, if you don't want to do it, so be it

kylo252 commented on utils/installer/install.sh:
```diff
@@ -1,272 +1,271 @@
+
+function __add_separator() {
+  local DIV_WIDTH="$1"
+  printf "%${DIV_WIDTH}s\n" ' ' | tr ' ' -
```
It's so that I don't have to escape the quote marks since the variable name would become `DIV_WIDTHs`, notice the `s`.

Tastyep commented on utils/installer/install.sh:
```diff
@@ -1,272 +1,280 @@
-  sudo apt install npm
-}
+#Set branch to master unless specified by the user
+declare -r LVBRANCH="${LVBRANCH:-master}"
```
`declare -r LVBRANCH=${LVBRANCH:-"master"}`

kylo252 replied:
That's not necessary. I needed to change it to rolling anyway.

Tastyep replied:
I'd like that you quote the inner most variable instead of outer so that is also matches other lines

Tastyep commented on utils/installer/install.sh:
```diff
@@ -1,272 +1,280 @@
+    __attempt_to_install_with_cargo "fd-find"
+  fi
+  if ! command -v rg &>/dev/null; then
+    __attempt_to_install_with_cargo "ripgrpe"
```
ripgrep

Tastyep commented on utils/installer/install.sh:
```diff
@@ -1,272 +1,280 @@
-    mkdir -p "$HOME/.local/share/applications"
-    cp "$HOME/.local/share/lunarvim/lvim/utils/desktop/lvim.desktop" "$HOME/.local/share/applications/lvim.desktop"
+function setup_lvim() {
+
```
blank line

abzcoding commented:
this doesn't work in mac os 😢 
<img width="367" alt="Screen Shot 2021-08-22 at 8 49 07 PM" src="https://user-images.githubusercontent.com/10992695/130362438-e3fffe42-1a86-4985-bf38-d7ba23ec7b89.png">

here is the difference between linux/osx
<img width="356" alt="Screen Shot 2021-08-22 at 8 52 53 PM" src="https://user-images.githubusercontent.com/10992695/130362545-710f8641-06ee-43bd-b4b3-1938a5e10c29.png">

kylo252 commented:
> this doesn't work in mac os 😢 
> <img width="367" alt="Screen Shot 2021-08-22 at 8 49 07 PM" src="https://user-images.githubusercontent.com/10992695/130362438-e3fffe42-1a86-4985-bf38-d7ba23ec7b89.png">
> 
> here is the difference between linux/osx
> <img width="356" alt="Screen Shot 2021-08-22 at 8 52 53 PM" src="https://user-images.githubusercontent.com/10992695/130362545-710f8641-06ee-43bd-b4b3-1938a5e10c29.png">
> 

That's what I tried to avoid originally.
@Tastyep, any ideas/workarounds?

abzcoding commented on utils/installer/install.sh:
```diff
@@ -1,272 +1,287 @@
-  sudo apt install nodejs
-  sudo apt install npm
-}
+#Set branch to master unless specified by the user
```
can we update this comment as well :)

kylo252 replied:
oops :)

It cannot be safely used for master until https://github.com/LunarVim/LunarVim/pull/1354 is there.

abzcoding commented:
btw can we have a warning or sth that mac users need to have `$HOME/.local/` in their path or that the `lvim` binary is located at `~/.local/lvim` and let users know to move it somewhere else if they want for example
```sh
sudo mv ~/.local/lvim /usr/local/bin/lvim
```

abzcoding reviewed (approved):
LGTM 🚀
